File: src/explainable_engine/models/request.py
from __future__ import annotations

import warnings

from pydantic import BaseModel, Field, model_validator


class Component(BaseModel):
    id: str | None = None
    name: str
    value: float
    weight: float = Field(ge=0.0, le=1.0)
    confidence: float = Field(ge=0.0, le=1.0, default=1.0)
    components: list[Component] | None = None

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "name": "trend_strength",
                    "value": 0.8,
                    "weight": 0.4,
                    "confidence": 0.9,
                }
            ]
        }
    }


class ExplainOptions(BaseModel):
    include_graph: bool = True
    include_drivers: bool = True
    max_drivers: int = Field(default=5, ge=1)
    max_depth: int = Field(default=10, ge=1)


class ExplainRequest(BaseModel):
    target: str
    value: float
    components: list[Component]
    options: ExplainOptions = Field(default_factory=ExplainOptions)
    metadata: dict | None = None

    @model_validator(mode="after")
    def warn_if_weights_exceed_one(self) -> ExplainRequest:
        total = sum(c.weight for c in self.components)
        if total > 1.0 + 1e-9:
            warnings.warn(
                f"Component weights sum to {total:.4f}, which exceeds 1.0",
                stacklevel=2,
            )
        return self

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "target": "market_regime_score",
                    "value": 0.72,
                    "components": [
                        {
                            "name": "trend_strength",
                            "value": 0.8,
                            "weight": 0.4,
                            "confidence": 0.9,
                        },
                        {
                            "name": "volatility",
                            "value": 0.5,
                            "weight": 0.3,
                            "confidence": 0.7,
                        },
                        {
                            "name": "momentum",
                            "value": 0.6,
                            "weight": 0.3,
                            "confidence": 0.85,
                        },
                    ],
                }
            ]
        }
    }

File: internal/models/request.go
package models

// Component represents an input component that contributes to the target value.
type Component struct {
	ID         string      `json:"id,omitempty"`
	Name       string      `json:"name"`
	Value      float64     `json:"value"`
	Weight     float64     `json:"weight"`
	Confidence float64     `json:"confidence"`
	Components []Component `json:"components,omitempty"`
}

// ExplainOptions configures what to include in the explanation response.
type ExplainOptions struct {
	IncludeGraph   bool `json:"include_graph"`
	IncludeDrivers bool `json:"include_drivers"`
	MaxDrivers     int  `json:"max_drivers"`
	MaxDepth       int  `json:"max_depth"`
}

// DefaultExplainOptions returns sensible defaults.
func DefaultExplainOptions() ExplainOptions {
	return ExplainOptions{
		IncludeGraph:   true,
		IncludeDrivers: true,
		MaxDrivers:     5,
		MaxDepth:       10,
	}
}

// ExplainRequest is the input payload for creating an explanation.
type ExplainRequest struct {
	Target     string            `json:"target"`
	Value      float64           `json:"value"`
	Components []Component       `json:"components"`
	Options    *ExplainOptions   `json:"options,omitempty"`
	Metadata   map[string]string `json:"metadata,omitempty"`
}

// GetOptions returns the options, using defaults if not specified.
func (r *ExplainRequest) GetOptions() ExplainOptions {
	if r.Options != nil {
		opts := *r.Options
		if opts.MaxDrivers == 0 {
			opts.MaxDrivers = 5
		}
		if opts.MaxDepth == 0 {
			opts.MaxDepth = 10
		}
		return opts
	}
	return DefaultExplainOptions()
}
